src/tokenserver/auth/browserid.rs

```diff
@@ -0,0 +1,469 @@
+use core::time::Duration;
+
+use async_trait::async_trait;
+use reqwest::{blocking::Client as ReqwestClient, StatusCode};
+use serde::{de::Deserializer, Deserialize, Serialize};
+
+use super::VerifyToken;
+use crate::tokenserver::{
+    error::{ErrorLocation, TokenserverError},
+    settings::Settings,
+};
+
+/// The information extracted from a valid BrowserID assertion.
+#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
+pub struct VerifyOutput {
+    pub device_id: Option<String>,
+    pub email: String,
+    pub generation: Option<i64>,
+    pub keys_changed_at: Option<i64>,
+}
+
+/// The verifier used to verify BrowserID assertions.
+#[derive(Clone)]
+pub struct RemoteVerifier {
+    audience: String,
+    issuer: String,
+    fxa_verifier_url: String,
+    request_client: ReqwestClient,
+}
+
+impl From<&Settings> for RemoteVerifier {
+    fn from(settings: &Settings) -> Self {
+        Self {
+            audience: settings.fxa_browserid_audience.clone(),
+            issuer: settings.fxa_browserid_issuer.clone(),
+            request_client: ReqwestClient::builder()
+                .timeout(Duration::new(settings.fxa_browserid_request_timeout, 0))
+                .build()
+                .expect("Couldn't initialize BrowserID verifier request client"),
+            fxa_verifier_url: settings.fxa_browserid_server_url.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl VerifyToken for RemoteVerifier {
+    type Output = VerifyOutput;
+
+    /// Verifies a BrowserID assertion. Returns `VerifyOutput` for valid assertions and a
+    /// `TokenserverError` for invalid assertions.
+    async fn verify(&self, assertion: String) -> Result<VerifyOutput, TokenserverError> {
```

Comment:
This is comparable to [this method](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/verifiers.py#L137) on the old Tokenserver. Note that the Python function throws a number of different exceptions, which are turned into errors reported to the client [here](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/views.py#L180)